### PR TITLE
feat!: Oauth register/login implementation

### PR DESCRIPTION
**Pull request by jonas-martinez**
Closes https://github.com/lenra-io/client-common/issues/82

## Checklist
- [x] I didn't over-scope my PR
- [x] My PR title matches the [commit convention](https://www.conventionalcommits.org/en/v1.0.0/)
- [ ] I did not include breaking changes
- [x] I made my own code-review before requesting one

### I included unit tests that cover my changes
  - [ ] 👍 yes
  - [x] 🙅 no, because they aren't needed
  - [ ] 🙋 no, because I need help

### I added/updated the documentation about my changes

- [ ] 📜 README.md
- [ ] 📕 docs/*.md
- [ ] 📓 docs.lenra.io
- [x] 🙅 no documentation needed

### CODE REVIEW

File: lib/views/auth/oauth_page.dart [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+                CguModel cguModel = context.read<CguModel>();
+
+                // Accept latest CGU if not already accepted
+                // This is done by default when the user registers as long as the oauth does not implement it
+                if (!(await cguModel.userAcceptedLatestCgu()).accepted) {
+                  CguResponse latestCGUResponse = await cguModel.getLatestCgu();
+                  await cguModel.acceptCgu(latestCGUResponse.id);
+                }
+
+                // TODO: Update the role of the user
```
Comment (jonas-martinez) [inline lib/views/auth/oauth_page.dart:42]: @taorepoara @Nesqwik How do we want to update the role of the user with the oauth system? Should it be done in the consent process of oauth (CGUs could differ between client and backoffice) or should it be done in the backoffice by maybe keeping the validation dev page?

**Comment by jonas-martinez** [comment]
I still need to find a way to check if the user is authenticated without running the `authenticate` function that opens the popup.

**Comment by jonas-martinez** [comment]
When I try to connect to the oauth with a bad access_token (from another hydra client for example), I get this error
 ```
Request: GET /api/me/opened_apps
** (exit) an exception was raised:
    ** (UndefinedFunctionError) function nil.id/0 is undefined
        nil.id()
        (lenra_web 0.1.0) lib/lenra_web/plug/verify_cgu.ex:14: LenraWeb.Plug.VerifyCgu.call/2
        (lenra_web 0.1.0) LenraWeb.Router.ensure_cgu_accepted/2
        (lenra_web 0.1.0) lib/lenra_web/router.ex:1:
...
```
 
The `current_resource` method returns `nil` for the user

**Comment by Nesqwik** [comment]
@jonas-martinez how is it going ? Do you need help ? :)

**Comment by jonas-martinez** [comment]
@Nesqwik I'm doing a lot of rework because there was too much problems with the error handling. I'm working on another issue for the moment but I expect to finish this feature before the end of the week.

File: lib/navigator/guard.dart [hunk 2]
```diff
@@ -3,110 +3,48 @@ import 'dart:async';
...
 import 'package:client_common/navigator/common_navigator.dart';
+import 'package:client_common/oauth/oauth_model.dart';
+import 'package:client_common/views/auth/oauth_page.dart';
 import 'package:flutter/material.dart';
 import 'package:provider/provider.dart';
 
 /// This class defines guards that are used to stop the user from accessing certain pages.
 class Guard {
-  Future<bool> Function(BuildContext) isValid;
+  Future<bool> Function(BuildContext, Map<String, dynamic>?) isValid;
```
Comment (taorepoara) [inline lib/navigator/guard.dart:15]: What is the metadata param  made for ?

It does not seems to be used...

**Review by taorepoara — CHANGES_REQUESTED** [review]
Some little comments.

File: lib/navigator/common_navigator.dart [hunk 3]
```diff
@@ -21,81 +17,34 @@ class CommonNavigator {
   static GoRoute changeLostPassword = GoRoute(
```
[the remark this answers is not in the document]
Comment (jonas-martinez) [reply lib/navigator/common_navigator.dart]: FIxed it

**Review by taorepoara — APPROVED** [review]

**Pull request merged**